libraries/AP_Scripting/examples/TerrainDetector.lua

```diff
@@ -0,0 +1,244 @@
+
+-- Yuri -- Apr 2022
+
+LICENSE - GNU GPLv3 https://www.gnu.org/licenses/gpl-3.0.en.html
+
+Concept first presented during this live stream:
+https://www.youtube.com/watch?v=UdXGXjigxAo&t=7155s
+Credit to @ktrussell for the idea and discussion!
+------------------------------------------------------------------------------]]
+
+local SCRIPT_NAME     = 'TerrainDetector'
+local RUN_INTERVAL_MS =  25  -- needs to be pretty fast for good detection
+local SBY_INTERVAL_MS = 500  -- slower interval when detection is disabled
+
+-- tuning values
+local SCR_USER_SPEED_PARAM     = 'SCR_USER1'  -- (m/s) set rough terrain speed here
+                                              -- set to -1 to disable detection
+local IMPULSE_GZ_THRESHOLD     = 1.33  -- (G) slow down > this Gz impulse
+local GYRO_RATE_THRESHOLD      =   28  -- (deg/s) slow down > this rate transient
+local ROUGH_TERRAIN_TIMEOUT_MS = 7500  -- min time to remain slowed down for rough terrain
+
+-- gain values are used for smooth terrain detection
+-- set less than 1.0 to be more restrictive about resuming normal speed
+local IMPULSE_GAIN = 0.9
+local GYRO_GAIN    = 0.8
+
+-- GCS messages
+local VERBOSE_MODE = 2 -- 0 to suppress all GCS messages,
+                       -- 1 for normal status messages
+                       -- 2 for additional debug messages
+local MSG_NORMAL = 1
+local MSG_DEBUG  = 2
+
+-- MAVLink values
+local CMD_DO_CHANGE_SPEED  = 178
+local MAV_SEVERITY_WARNING =   4
+local MAV_SEVERITY_INFO    =   6
+
+-- mathematical/physical constants
+local G          = -9.81 -- m/s/s
+local RAD_TO_DEG = 180.0 / math.pi
+
+-- wrapper for gcs:send_text()
+local function gcs_msg(msg_type, severity, txt)
+    if type(msg_type) == 'string' then
+    -- allow just a string to be passed for simple/routine messages
+        txt      = msg_type
+        msg_type = MSG_NORMAL
+        severity = MAV_SEVERITY_INFO
+    end
+    if msg_type <= VERBOSE_MODE then
+        gcs:send_text(severity, string.format('%s: %s', SCRIPT_NAME, txt))
+    end
+end
+
+local last_g_z   = 0  -- to calculate impulse Gz
+local timeout_ms = 0
+
+-- temp variables for restoring waypoint changes
+local temp_wp       = nil
+local temp_wp_index = nil
+
+-- debug values
+local max_g_z       = 0
+local max_gyro_rate = 0
+
+-- waypoint template for speed changes
+local wp_do_change_speed = mavlink_mission_item_int_t()
+wp_do_change_speed:command(CMD_DO_CHANGE_SPEED)
+
+local wp_speed_normal = param:get('WP_SPEED')
+local wp_speed_rough  = param:get(SCR_USER_SPEED_PARAM)
+
+local function insert_do_change_speed(index)
+    local curr_mission = {}
+    for i = 1, mission:num_commands() do
+        curr_mission[i] = mission:get_item(i)
+    end
+    wp_do_change_speed:param2(wp_speed_normal)
+    table.insert(curr_mission, index, wp_do_change_speed)
+    for i, item in pairs(curr_mission) do
+        mission:set_item(i, item)
+    end
+end
+
+local function do_change_speed(speed)

```

Comment:
I would be tempted to wait for a direct binding for this one, what do you think?

https://github.com/ArduPilot/ardupilot/issues/20618

Reply:
There may be value in the interim to showing this example with waypoint insertion and manipulation.  I do intend to update the script when the binding is added.

I'll leave it up to you whether to include the rather "hacky" example or wait.

Reply:
Removed the TerrainDetector example in favor of an upcoming PR to add the binding as suggested